DEV: Update default tagName and connectorTagName for plugin outlets

davidtaylorhq opened the pull request:
This commit should be a no-op for all existing core outlets. Outlets which are introduced by themes/plugins may see a change in behavior, and should follow the steps below if they want to maintain their previous behavior.

`tagName="" connectorTagName=""` is almost always the correct choice for plugin outlets. 40eba8c introduced a `noTags=true` shortcut which achieved this, and left a comment saying it should be the future default. This commit does exactly that. To avoid any breaking changes for plugins, all existing plugin outlets have been reviewed and adjusted by following this logic:

1) If `noTags=true`, remove the `noTags` parameter, and do not complete any further steps
2) If `tagName` is not specified, set `tagName="span"` (the previous default)
3) If `connectorTagName` is not specified, set `selectorTagName="div"` (the previous default)
4) If `tagName=""`, remove it
5) If `connectorTagName=""`, remove it

The updates were accomplished with the help of a ruby script:

```ruby
def removeAttr(tag, attribute)
  tag = tag.sub /\s#{attribute}="?\w*"? /, " "
  tag = tag.sub /\s#{attribute}="?\w*"?}}/, "}}"
  tag = tag.sub /^\s*#{attribute}="?\w*"?\n/, ""
  tag
end

files = Dir.glob("app/assets/javascripts/**/*.hbs")

puts "Checking #{files.count} files..."
files.each do |f|
  content = File.read(f)

  count = 0
  edits = 0
  content.gsub!(/{{\s*plugin-outlet.*?}}/m) do |match|
    count += 1

    result = match

    noTags = result.include?("noTags=true")
    tagName = result[/tagName="(\w*)"/, 1]
    connectorTagName = result[/connectorTagName="(\w*)"/, 1]

    if noTags
      result = removeAttr(result, "noTags")
    else
      if connectorTagName == ""
        result = removeAttr(result, "connectorTagName")
      elsif connectorTagName.nil?
        result = result.sub(/name="[\w-]+"/) { |m| "#{m} connectorTagName=\"div\"" }
      end

      if tagName == ""
        result = removeAttr(result, "tagName")
      elsif tagName.nil?
        result = result.sub(/name="[\w-]+"/) { |m| "#{m} tagName=\"span\"" }
      end
    end

    edits += 1 if match != result

    result
  end

  puts "#{count} outlets, #{edits} edited -> #{f}"

  File.write(f, content)
end
```

CvX reviewed (commented):
TIL, there was `noTags` option!

A thing that should be fixed (not necessarily in this PR) are all instances of `tagName="span" connectorTagName="div"` (i.e. the previous default) since you can't put divs inside spans.

> If `connectorTagName` is not specified, set `selectorTagName="div"` (the previous default)

A typo, should be `connectorTagName` in both instances.

eviltrout reviewed (approved):
This is quite nice and I like the change. I kinda wish we always defaulted to `noTags` because there are a lot of extra divs we never need, but this is a great step in the right direction.

SamSaffron commented:
many conflicts, what shall we do with this PR @davidtaylorhq  ?

davidtaylorhq commented:
I need to rebase and do another pass for new plugin outlets. Will try and get to it this week

K-adot commented on app/assets/javascripts/discourse/app/templates/composer.hbs:
```diff
@@ -193,11 +193,11 @@
+            {{plugin-outlet name="composer-mobile-buttons-bottom" tagName="span" args=(hash model=model)}}
 
             {{#if allowUpload}}
               <a id="mobile-file-upload" class="btn btn-default no-text mobile-file-upload {{if isUploading "hidden"}}"  aria-label={{i18n "composer.upload_title"}}>
```
membantu

K-adot reviewed (commented):
update

davidtaylorhq commented:
I've reworked this on top of the latest `main`, and added a more verbose commit message / PR description. (this time I did the edits with the help of a script - I couldn't bring myself to manually edit almost 200 plugin outlets for a second time! 😅)

eviltrout reviewed (approved):
Let's do it before the next release.
